lib/mastodon/redis_configuration.rb

```diff
@@ -55,42 +54,53 @@ def cache_namespace
     namespace ? "#{namespace}_cache" : 'cache'
   end
 
-  def setup_base_redis_url
-    url = ENV.fetch('REDIS_URL', nil)
-    return url if url.present?
-
-    user     = ENV.fetch('REDIS_USER', '')
-    password = ENV.fetch('REDIS_PASSWORD', '')
-    host     = ENV.fetch('REDIS_HOST', 'localhost')
-    port     = ENV.fetch('REDIS_PORT', 6379)
-    db       = ENV.fetch('REDIS_DB', 0)
+  def setup_config(prefix: nil, defaults: {})
+    prefix = "#{prefix}REDIS_"
 
-    construct_uri(host, port, db, user, password)
-  end
+    url       = ENV.fetch("#{prefix}URL", nil)
+    user      = ENV.fetch("#{prefix}USER", nil)
+    password  = ENV.fetch("#{prefix}PASSWORD", nil)
+    host      = ENV.fetch("#{prefix}HOST", defaults[:host])
+    port      = ENV.fetch("#{prefix}PORT", defaults[:port])
+    db        = ENV.fetch("#{prefix}DB", defaults[:db])
+    name      = ENV.fetch("#{prefix}SENTINEL_MASTER", nil)
+    sentinels = parse_sentinels(ENV.fetch("#{prefix}SENTINELS", nil))
 
-  def setup_prefixed_redis_url(prefix)
-    prefix = "#{prefix.to_s.upcase}_"
-    url = ENV.fetch("#{prefix}REDIS_URL", nil)
+    return { url:, driver: } if url
 
-    return url if url.present?
+    if name.present? && sentinels.present?
+      host = name
+      port = nil
+      db ||= 0
+    else
+      sentinels = nil
+    end
 
-    user     = ENV.fetch("#{prefix}REDIS_USER", nil)
-    password = ENV.fetch("#{prefix}REDIS_PASSWORD", nil)
-    host     = ENV.fetch("#{prefix}REDIS_HOST", nil)
-    port     = ENV.fetch("#{prefix}REDIS_PORT", nil)
-    db       = ENV.fetch("#{prefix}REDIS_DB", nil)
+    url = construct_uri(host, port, db, user, password)
 
-    if host.nil?
-      base[:url]
+    if url.present?
+      { url:, driver:, name:, sentinels: }
     else
-      construct_uri(host, port, db, user, password)
+      # Fall back to base config. This has defaults for the URL
+      # so this cannot lead to an endless loop.
+      base
     end
   end
 
   def construct_uri(host, port, db, user, password)
+    return nil if host.blank?
+
     Addressable::URI.parse("redis://#{host}:#{port}/#{db}").tap do |uri|
       uri.user = user if user.present?
       uri.password = password if password.present?
     end.normalize.to_str
   end
+
+  def parse_sentinels(sentinels_string)
+    (sentinels_string || '').split(',').map do |sentinel|
+      host, port = sentinel.split(':')
+      port = port.present? ? port.to_i : 26_379
```

Comment:
I supported using SENTINEL_PORT or something too, we should probably make sure that's the same; I don't mind using a hardcoded default port

Reply:
I *knew* I missed something. Thank you so much for catching this.

I guess this is really helpful when all your sentinels use the same (non-standard) port.

Reply:
#31767